Yuval/lbm1 5419 add uncorr

yuval-lb opened the pull request:
We want to add uncorrectable feature to qemu-nvme.

1.  create bitmap to follow which block will raise NVME_UNRECOVERED_READ
2.  add NVME_CMD_WRITE_UNCOR - set range of blocks to raise "unrecovered" error
3.  add in nvme_rw - if its read check if the block is uncorrectable and raise error
                                - if it write - write reset the block to be correctable.

I am not sure this is enough to work.
@roys-lb , I need to add something to qemu/block/nvme.c?

maor-lb commented on hw/block/nvme.c:
```diff
@@ -1320,6 +1321,7 @@ static void nvme_realize(PCIDevice *pci_dev, Error **errp)
         id_ns->ncap  = id_ns->nuse = id_ns->nsze =
             cpu_to_le64(n->ns_size >>
                 id_ns->lbaf[NVME_ID_NS_FLBAS_INDEX(ns->id_ns.flbas)].ds);
+        id_ns->uncorrectable = bitmap_new(id_ns->nsze)
```
Looks good

maor-lb commented on hw/block/nvme.c:
```diff
@@ -49,6 +49,44 @@
+
+static int check_uncorrectable(NvmeNamespace *ns, is_write, slba, nlb)
+{
+    for(int i=0; i<=nlba; i++){
```
Isn't convention in qemu is to have space around <= ?

maor-lb commented on hw/block/nvme.c:
```diff
@@ -49,6 +49,44 @@
+    uint32_t nlb  = le32_to_cpu(cmd->nlb) + 1;
+    uint64_t slba = le64_to_cpu(cmd->slba);
+    if (unlikely((slba + nlb) > ns->id_ns.nsze)) {
+    block_acct_invalid(blk_get_stats(n->conf.blk), acct);
```
missing indent here

maor-lb commented on hw/block/nvme.c:
```diff
@@ -49,6 +49,44 @@
             " in %s: " fmt "\n", __func__, ## __VA_ARGS__); \
     } while (0)
 
+
```
@yuval please add sing off to commit message, I think we should review the commit guidelines of Qemu project

maor-lb commented on hw/block/nvme.c:
```diff
@@ -49,6 +49,44 @@
+    }
+
+    for(int i=0; i<=nlba; i++){
+        ns->uncorrectable[slba+i]=1
```
@yuval-lb you should use bitmap api, this loop is wrong cause only single bit allocated per lba,e.g

void bitmap_set(unsigned long *map, long i, long len);
void bitmap_set_atomic(unsigned long *map, long i, long len);
void bitmap_clear(unsigned long *map, long start, long nr);
bool bitmap_test_and_clear_atomic(unsigned long *map, long start, long nr);
void bitmap_copy_and_clear_atomic(unsigned long *dst, unsigned long *src,
                                  long nr);

yuval-lb replied:
+1

maor-lb commented on hw/block/nvme.c:
```diff
@@ -49,6 +49,44 @@
+{
+    for(int i=0; i<=nlba; i++){
+        if(ns->uncorrectable[slba+i]==1){
+            ns->uncorrectable[slba+i]=0
```
@yuval-lb same of above for bit API

yuval-lb replied:
+1

maor-lb reviewed (commented):
@yuval-lb see some comments inside

Commit:
nvme.c uncorrectable logic

qemu/hw/block/nvme.c:
1. add NVMe cmd option - NVME_CMD_WRITE_UNCOR
2. set_uncorrectable_blk - set uncorrectable blocks
3. check_and_update_uncorrectable - in read and write we need to check
   uncorrectable - case of hit in read then we return NVME_UNRECOVERED_READ
   case of hit in write we set the bit to 1 - as write reset uncorrectable.

Signed-off-by: yuval brave <yuval@lightbitslab.com>

Commit:
add uncorrectalbe bitmap to NVMeNamespace

qemu/hw/block/nvme.c: create bitmap for uncorrectable
qemu/hw/block/nvme.h: add uncorrectable bitmap pointer to NVMeNamespace

Signed-off-by: yuval brave <yuval@lightbitslab.com>

yuval-lb commented:
@maor-lb I updated the code, but must say its a "shot in the dark" as I have not tested it yet.
